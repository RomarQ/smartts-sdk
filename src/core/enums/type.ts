export enum TypeAtom {
    record = 'record',
    variant = 'variant',
    chain_id = 'chain_id',
    unit = 'unit',
    nat = 'nat',
    int = 'int',
    mutez = 'mutez',
    bool = 'bool',
    bytes = 'bytes',
    string = 'string',
    address = 'address',
    timestamp = 'timestamp',
    bls12_381_fr = 'bls12_381_fr',
    bls12_381_g1 = 'bls12_381_g1',
    bls12_381_g2 = 'bls12_381_g2',
    key = 'key',
    key_hash = 'key_hash',
    list = 'list',
    set = 'set',
    option = 'option',
    map = 'map',
    big_map = 'bigmap',
    signature = 'signature',
    operation = 'operation',
    never = 'never',
    tuple = 'tuple',
    lambda = 'lambda',
    ticket = 'ticket',
    sapling_state = 'sapling_state',
    contract = 'contract',
    sapling_transaction = 'sapling_transaction',
}

export default TypeAtom;
