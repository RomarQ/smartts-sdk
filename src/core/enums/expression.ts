export enum ExpressionAtom {
    // Math
    add = 'add',
    mul = 'mul',
    sub = 'sub',
    truediv = 'truediv',
    // Comparison
    eq = 'eq',
    neq = 'neq',
    lt = 'lt',
    gt = 'gt',
    le = 'le',
    ge = 'ge',
    //
    items = 'items',
    update_map = 'update_map',
    getItem = 'getItem',
    getItemDefault = 'getItemDefault',
    getItemMessage = 'getItemMessage',
    attr = 'attr',
    params = 'params',
    getLocal = 'getLocal',
    iter = 'iter',
    // Blockchain property
    amount = 'amount',
    balance = 'balance',
    chain_id = 'chain_id',
    level = 'level',
    now = 'now',
    self = 'self',
    self_address = 'self_address',
    sender = 'sender',
    source = 'source',
    total_voting_power = 'total_voting_power',
    voting_power = 'voting_power',
    // Operation
    operations = 'operations',
    transfer = 'transfer',
    set_delegate = 'set_delegate',
    create_contract = 'create_contract',
    // List instructions
    cons = 'cons',
    //
    contract = 'contract',
    // Variant
    openVariant = 'openVariant',
    isVariant = 'isVariant',
    cases_arg = 'cases_arg',
    variant_arg = 'variant_arg',
}

export default ExpressionAtom;
