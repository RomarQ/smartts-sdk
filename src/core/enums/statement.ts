export enum StatementAtom {
    verify = 'verify',
    ifBlock = 'ifBlock',
    elseBlock = 'elseBlock',
    forGroup = 'forGroup',
    whileBlock = 'whileBlock',
    set = 'set',
    defineLocal = 'defineLocal',
    result = 'result',
    failwith = 'failwith',
    match = 'match',
    match_cases = 'match_cases',
    set_type = 'set_type',
}

export default StatementAtom;
