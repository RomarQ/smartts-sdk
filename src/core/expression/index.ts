export * from './literal';
export * from './expression';
export * from './logic';
