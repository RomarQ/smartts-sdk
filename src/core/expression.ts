import type { IExpression } from '../typings/expression';
import type { ILiteral } from '../typings/literal';
import type { IStatement } from '../typings/statement';
import type { IType } from '../typings/type';
import { GetProperty, LambdaArgument } from '../expression/variables';
import { capitalizeBoolean, LineInfo, parenthesis } from '../misc/utils';
import ValueAtom from './enums/literal';
import TypeAtom from './enums/type';

export class Expression<T extends ValueAtom> implements IExpression<T> {
    _isExpression = true as const;
    // Used for type checking
    _type = {} as T;
    private args;

    constructor(public name: string, ...args: (IExpression | LineInfo | IType | string)[]) {
        this.args = args || [];
    }

    static proxyHandler: ProxyHandler<IExpression> = {
        get: function (target, prop: string, receiver) {
            if (prop in target || typeof prop === 'symbol') {
                return Reflect.get(target, prop, receiver);
            }
            return GetProperty(target, prop);
        },
        apply: function (target, thisArg, argumentsList) {
            console.log(target, thisArg, argumentsList);
            return target;
        },
    };

    toString() {
        return `(${[this.name, ...this.args].join(' ')})`;
    }
}

export class LiteralExpression<T extends ValueAtom> implements ILiteral<T> {
    _isExpression = true as const;
    // Used for type checking
    _type = {} as T;

    constructor(private name: T, private values: (number | string | boolean | IExpression)[], private line: LineInfo) {}

    toString() {
        switch (this.name) {
            case ValueAtom.unit:
                return parenthesis(this.name);
            case ValueAtom.nat:
            case ValueAtom.int:
            case ValueAtom.mutez:
            case ValueAtom.bytes:
            case ValueAtom.bool:
            case ValueAtom.string:
            case ValueAtom.chain_id_cst:
            case ValueAtom.timestamp:
            case ValueAtom.bls12_381_fr:
            case ValueAtom.bls12_381_g1:
            case ValueAtom.bls12_381_g2:
            case ValueAtom.key:
            case ValueAtom.key_hash:
            case ValueAtom.signature:
            case ValueAtom.address:
                return `(literal (${this.name} ${this.values.join(' ')}) ${this.line})`;
            case ValueAtom.list:
            case ValueAtom.set:
            case ValueAtom.tuple:
                return `(${this.name} ${this.line} ${this.values.join(' ')})`;
            case ValueAtom.Some:
            case ValueAtom.None:
                return `(${ValueAtom.variant} "${this.name}" ${this.values.join(' ') || `(${ValueAtom.unit})`} ${
                    this.line
                })`;
            default:
                return `(${this.name} ${this.values.join(' ')} ${this.line})`;
        }
    }
}

export class RecordLiteral implements ILiteral<ValueAtom.record> {
    _isExpression = true as const;
    // Used for type checking
    _type = {} as ValueAtom.record;

    constructor(private fields: Record<string, IExpression>, private line: LineInfo) {}

    private buildFields = (fields: Record<string, IExpression>) => {
        return Object.entries(fields).map(([field, expr]) => `(${field} ${expr.toString()})`);
    };

    toString(): string {
        return `(${TypeAtom.record} ${this.line} ${this.buildFields(this.fields).join(' ')})`;
    }
}

export class MapLiteral<T extends ValueAtom.map | ValueAtom.big_map> implements ILiteral<T> {
    _isExpression = true as const;
    _type = {} as T;
    type = {} as IType;

    constructor(
        private prim: T,
        private rows: IExpression[][],
        keyType: IType,
        valueType: IType,
        private line: LineInfo,
    ) {}

    private buildEntry = ([key, value]: IExpression[]) => {
        return `(${key.toString()} ${value.toString()})`;
    };

    toString() {
        return `(${this.prim} ${this.line} ${this.rows.map(this.buildEntry).join(' ')})`;
    }
}

export class LambdaLiteral implements ILiteral<ValueAtom.lambda> {
    _isExpression = true as const;
    // Used for type checking
    _type = ValueAtom.lambda as const;
    type = {} as unknown as IType;
    static idCounter = 0;
    private identifier: number;
    private withStorage?: 'read-write' | 'read-only';
    private withOperations = false;

    constructor(
        private statements: IStatement[],
        private inType: IType,
        private argumentName: string,
        private line: LineInfo,
    ) {
        this.identifier = ++LambdaLiteral.idCounter;
    }

    public get id(): Readonly<number> {
        return this.identifier;
    }

    public setInputType(type: IType) {
        this.inType = type;
        return this;
    }

    public code(buildStatements: (arg: IExpression) => IStatement[]) {
        const param = LambdaArgument(this.argumentName, this.inType, this.id, this.line);
        this.statements = buildStatements(param);
        return this;
    }

    toString() {
        return `(${ValueAtom.lambda} ${this.id} ${this.withStorage || 'None'} ${capitalizeBoolean(
            this.withOperations,
        )} "${this.argumentName}" ${this.line} (${this.statements.join(' ')}))`;
    }
}
