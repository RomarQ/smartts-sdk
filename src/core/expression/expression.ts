import { proxy } from '../../misc/proxy';
import { LineInfo } from '../../misc/utils';
import { IExpression } from '../../typings/expression';
import { ILiteral } from '../../typings/literal';
import { IType } from '../../typings/type';

export class Expression implements IExpression {
    _isExpression = true as const;
    args;

    constructor(public name: string, ...args: (IExpression | LineInfo | IType | string | ILiteral<unknown>)[]) {
        this.args = args || [];
    }

    static proxyHandler: ProxyHandler<IExpression> = {
        get: function (target, prop: string, receiver) {
            if (prop in target || typeof prop === 'symbol') {
                return Reflect.get(target, prop, receiver);
            }
            return Expression.getAttr(target, prop);
        },
        apply: function (target, thisArg, argumentsList) {
            console.log(target, thisArg, argumentsList);
            return target;
        },
    };

    static getAttr = (from: IExpression, attr: string, line = new LineInfo()) =>
        proxy(new Expression('attr', from, `"${attr}"`, line), Expression.proxyHandler);

    toString() {
        return `(${[this.name, ...this.args].join(' ')})`;
    }
}
