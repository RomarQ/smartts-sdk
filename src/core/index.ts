import type { IToString, IToType } from '../typings';
import type { ILiteral } from '../typings/literal';
import type { IType } from '../typings/type';

import { Expression } from './expression';
import { Mutez } from './literal';
import Utils, { LineInfo } from '../misc/utils';
import { TUnit } from './type';

export class Flag {
    args: string[] = [];

    constructor(public flag: string, ...args: string[]) {
        this.args = args;
    }

    toString() {
        return `(${this.flag}${(' ' + this.args.join(' ')).trim()})`;
    }
}

interface EntryPointOptions {
    mock?: boolean;
    lazy?: boolean;
    lazyAndCodeless?: boolean;
}
export class EntryPoint {
    options = {
        mock: false,
        lazy: false,
        lazyAndCodeless: false,
    };
    inType?: IType;
    statements: IToString[] = [];

    constructor(public name: string, public line = new LineInfo()) {}

    config(options?: EntryPointOptions) {
        if (options?.mock) {
            this.options.mock = options.mock;
        }
        if (options?.lazy) {
            this.options.lazy = options.lazy;
        }
        if (options?.lazyAndCodeless) {
            this.options.lazyAndCodeless = options.lazyAndCodeless;
        }
        return this;
    }

    inputType(type: IType) {
        this.inType = type;
        return this;
    }

    code(callback: (arg: IToString) => IToString[]) {
        this.statements = callback(new Expression('params', new LineInfo()));
        return this;
    }

    toString() {
        const notMock = Utils.capitalizeBoolean(!this.options.mock);
        const isLazy = Utils.capitalizeBoolean(this.options.lazy);
        const isLazyAndCodeless = Utils.capitalizeBoolean(this.options.lazyAndCodeless);
        const hasParams = Utils.capitalizeBoolean(!!this.inType);
        return `(${this.name} ${notMock} ${isLazy} ${isLazyAndCodeless} ${hasParams} ${
            this.line
        } (${this.statements.join(' ')}))`;
    }
}

export class Contract {
    options: {
        initialBalance: ILiteral;
        flags: Flag[];
    } = {
        initialBalance: Mutez(0),
        flags: [],
    };
    public initialStorage: IToString & IToType;
    public entries: EntryPoint[];

    constructor(
        args: {
            initialStorage: IToString & IToType;
            entries: EntryPoint[];
        },
        public line = new LineInfo(),
    ) {
        this.initialStorage = args.initialStorage;
        this.entries = args.entries;
    }

    config(options?: { initialBalance?: ILiteral; flags?: Flag[] }) {
        if (options?.flags) {
            this.options.flags = options.flags;
        }
        if (options?.initialBalance) {
            this.options.initialBalance = options.initialBalance;
        }
        return this;
    }

    toString() {
        return `
        (
            template_id (static_id 0 ${this.line})
            storage ${this.initialStorage.toString()}
            storage_type (${this.initialStorage.toType()})
            messages (${this.entries.map((entry) => entry.toString()).join(' ')})
            flags (${this.options.flags.map((flag) => flag.toString())})
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance ${this.options.initialBalance}
        )
        `;
    }
}

export { default as Command } from './command';
export { default as Expression } from './expression';
export { default as Type } from './type';
export { default as Literal } from './literal';
export * from './blockchain_operations';
