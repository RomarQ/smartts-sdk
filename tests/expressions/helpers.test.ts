import { Contract, EntryPoint } from '../../src/core';
import { verifyContractCompilationOutput } from '../util';
import { SetValue } from '../../src/statement';
import { Address, ContractStorage, String } from '../../src/expression';
import { TBool } from '../../src/type';
import { AddressHelpers, StringHelpers } from '../../src/lib';

describe('Helpers', () => {
    it('StartsWith', () => {
        const contract = new Contract()
            .setStorageType(TBool())
            .addEntrypoint(
                new EntryPoint('ep1').code(() => [
                    SetValue(ContractStorage(), StringHelpers.StartsWith(String('Hello'), String('He'))),
                ]),
            );

        verifyContractCompilationOutput(contract);
    });
    it('EndsWith', () => {
        const contract = new Contract()
            .setStorageType(TBool())
            .addEntrypoint(
                new EntryPoint('ep1').code(() => [
                    SetValue(ContractStorage(), StringHelpers.EndsWith(String('Hello'), String('lo'))),
                ]),
            );

        verifyContractCompilationOutput(contract);
    });
    it('IsKT1', () => {
        const contract = new Contract()
            .setStorageType(TBool())
            .addEntrypoint(
                new EntryPoint('ep1').code(() => [
                    SetValue(ContractStorage(), AddressHelpers.IsKT1(Address('KT1R9M3MDffw7qSVSnbJs46aMC9YzzZz3aGT'))),
                ]),
            );

        verifyContractCompilationOutput(contract);
    });
});
