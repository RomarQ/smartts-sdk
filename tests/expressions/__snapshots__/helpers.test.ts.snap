// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helpers EndsWith 1`] = `
"
        (
            template_id (static_id 0 (\\"helpers.test.ts\\" 21))
            storage ()
            storage_type (\\"bool\\")
            messages ((ep1 True False False True (\\"helpers.test.ts\\" 24) ((set_type (params (\\"helpers.test.ts\\" 24)) \\"unit\\" (\\"helpers.test.ts\\" 29)) (set (data) (call_michelson (op \\"
        DUP;
        SIZE;
        DUP 3;
        SIZE;
        SWAP;
        PAIR;
        DUP;
        UNPAIR;
        COMPARE;
        GE;
        IF
        {
            UNPAIR;
            DUP 2;
            SWAP;
            SUB;
            ABS; # ABS is secure here because we already validated that (text length is greater or equal to postfix)
            SLICE;
            IF_NONE
            {
                DROP;
                PUSH bool False;
            }
            {
                COMPARE;
                EQ;
            };
        }
        {
            DROP 3;
            PUSH bool False;
        };
        \\" \\"string\\" \\"string\\" out \\"bool\\") (\\"string.ts\\" 95) (literal (string \\"Hello\\") (\\"helpers.test.ts\\" 24)) (literal (string \\"lo\\") (\\"helpers.test.ts\\" 24))) (\\"helpers.test.ts\\" 24)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"helpers.test.ts\\" 21))
        )
        "
`;

exports[`Helpers EndsWith 2`] = `
"{
    \\"micheline\\": \\"parameter (unit %ep1);\\\\nstorage   bool;\\\\ncode\\\\n  {\\\\n    CAR;\\\\n    PUSH string \\\\\\"lo\\\\\\";\\\\n    PUSH string \\\\\\"Hello\\\\\\";\\\\n    DUP;\\\\n    SIZE;\\\\n    DUP 3;\\\\n    SIZE;\\\\n    SWAP;\\\\n    PAIR;\\\\n    DUP;\\\\n    UNPAIR;\\\\n    COMPARE;\\\\n    GE;\\\\n    IF\\\\n      {\\\\n        DIG 3;\\\\n        DROP;\\\\n        UNPAIR;\\\\n        DUP 2;\\\\n        SWAP;\\\\n        SUB;\\\\n        ABS;\\\\n        SLICE;\\\\n        IF_NONE\\\\n          {\\\\n            DROP;\\\\n            PUSH bool False;\\\\n          }\\\\n          {\\\\n            COMPARE;\\\\n            EQ;\\\\n          };\\\\n      }\\\\n      {\\\\n        DROP 4;\\\\n        PUSH bool False;\\\\n      };\\\\n    NIL operation;\\\\n    PAIR;\\\\n  };\\",
    \\"json\\": [
        {
            \\"prim\\": \\"storage\\",
            \\"args\\": [
                {
                    \\"prim\\": \\"bool\\"
                }
            ]
        },
        {
            \\"prim\\": \\"parameter\\",
            \\"args\\": [
                {
                    \\"prim\\": \\"unit\\",
                    \\"annots\\": [
                        \\"%ep1\\"
                    ]
                }
            ]
        },
        {
            \\"prim\\": \\"code\\",
            \\"args\\": [
                [
                    {
                        \\"prim\\": \\"CAR\\"
                    },
                    {
                        \\"prim\\": \\"PUSH\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"string\\"
                            },
                            {
                                \\"string\\": \\"lo\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"PUSH\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"string\\"
                            },
                            {
                                \\"string\\": \\"Hello\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"DUP\\"
                    },
                    {
                        \\"prim\\": \\"SIZE\\"
                    },
                    {
                        \\"prim\\": \\"DUP\\",
                        \\"args\\": [
                            {
                                \\"int\\": \\"3\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"SIZE\\"
                    },
                    {
                        \\"prim\\": \\"SWAP\\"
                    },
                    {
                        \\"prim\\": \\"PAIR\\"
                    },
                    {
                        \\"prim\\": \\"DUP\\"
                    },
                    {
                        \\"prim\\": \\"UNPAIR\\"
                    },
                    {
                        \\"prim\\": \\"COMPARE\\"
                    },
                    {
                        \\"prim\\": \\"GE\\"
                    },
                    {
                        \\"prim\\": \\"IF\\",
                        \\"args\\": [
                            [
                                {
                                    \\"prim\\": \\"DIG\\",
                                    \\"args\\": [
                                        {
                                            \\"int\\": \\"3\\"
                                        }
                                    ]
                                },
                                {
                                    \\"prim\\": \\"DROP\\"
                                },
                                {
                                    \\"prim\\": \\"UNPAIR\\"
                                },
                                {
                                    \\"prim\\": \\"DUP\\",
                                    \\"args\\": [
                                        {
                                            \\"int\\": \\"2\\"
                                        }
                                    ]
                                },
                                {
                                    \\"prim\\": \\"SWAP\\"
                                },
                                {
                                    \\"prim\\": \\"SUB\\"
                                },
                                {
                                    \\"prim\\": \\"ABS\\"
                                },
                                {
                                    \\"prim\\": \\"SLICE\\"
                                },
                                {
                                    \\"prim\\": \\"IF_NONE\\",
                                    \\"args\\": [
                                        [
                                            {
                                                \\"prim\\": \\"DROP\\"
                                            },
                                            {
                                                \\"prim\\": \\"PUSH\\",
                                                \\"args\\": [
                                                    {
                                                        \\"prim\\": \\"bool\\"
                                                    },
                                                    {
                                                        \\"prim\\": \\"False\\"
                                                    }
                                                ]
                                            }
                                        ],
                                        [
                                            {
                                                \\"prim\\": \\"COMPARE\\"
                                            },
                                            {
                                                \\"prim\\": \\"EQ\\"
                                            }
                                        ]
                                    ]
                                }
                            ],
                            [
                                {
                                    \\"prim\\": \\"DROP\\",
                                    \\"args\\": [
                                        {
                                            \\"int\\": \\"4\\"
                                        }
                                    ]
                                },
                                {
                                    \\"prim\\": \\"PUSH\\",
                                    \\"args\\": [
                                        {
                                            \\"prim\\": \\"bool\\"
                                        },
                                        {
                                            \\"prim\\": \\"False\\"
                                        }
                                    ]
                                }
                            ]
                        ]
                    },
                    {
                        \\"prim\\": \\"NIL\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"operation\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"PAIR\\"
                    }
                ]
            ]
        }
    ],
    \\"smartpy\\": \\"import smartpy as sp\\\\n\\\\nclass Contract(sp.Contract):\\\\n  def __init__(self):\\\\n    self.init_type(sp.TBool)\\\\n\\\\n  @sp.entry_point\\\\n  def ep1(self, params):\\\\n    sp.set_type(params, sp.TUnit)\\\\n    self.data = sp.michelson(\\\\\\"\\\\\\\\n        DUP;\\\\\\\\n        SIZE;\\\\\\\\n        DUP 3;\\\\\\\\n        SIZE;\\\\\\\\n        SWAP;\\\\\\\\n        PAIR;\\\\\\\\n        DUP;\\\\\\\\n        UNPAIR;\\\\\\\\n        COMPARE;\\\\\\\\n        GE;\\\\\\\\n        IF\\\\\\\\n        {\\\\\\\\n            UNPAIR;\\\\\\\\n            DUP 2;\\\\\\\\n            SWAP;\\\\\\\\n            SUB;\\\\\\\\n            ABS; # ABS is secure here because we already validated that (text length is greater or equal to postfix)\\\\\\\\n            SLICE;\\\\\\\\n            IF_NONE\\\\\\\\n            {\\\\\\\\n                DROP;\\\\\\\\n                PUSH bool False;\\\\\\\\n            }\\\\\\\\n            {\\\\\\\\n                COMPARE;\\\\\\\\n                EQ;\\\\\\\\n            };\\\\\\\\n        }\\\\\\\\n        {\\\\\\\\n            DROP 3;\\\\\\\\n            PUSH bool False;\\\\\\\\n        };\\\\\\\\n        \\\\\\")('Hello', 'lo')\\\\n\\\\nsp.add_compilation_target(\\\\\\"visualtez_compilation\\\\\\", Contract())\\"
}"
`;

exports[`Helpers IsKT1 1`] = `
"
        (
            template_id (static_id 0 (\\"helpers.test.ts\\" 32))
            storage ()
            storage_type (\\"bool\\")
            messages ((ep1 True False False True (\\"helpers.test.ts\\" 35) ((set_type (params (\\"helpers.test.ts\\" 35)) \\"unit\\" (\\"helpers.test.ts\\" 40)) (set (data) (and (ge (literal (address KT1XvNYseNDJJ6Kw27qhSEDF8ys8JhDopzfG) (\\"address.ts\\" 15)) (literal (address KT1R9M3MDffw7qSVSnbJs46aMC9YzzZz3aGT) (\\"helpers.test.ts\\" 35)) (\\"address.ts\\" 15)) (le (literal (address KT18amZmM5W7qDWVt2pH6uj7sCEd3kbzLrHT) (\\"address.ts\\" 16)) (literal (address KT1R9M3MDffw7qSVSnbJs46aMC9YzzZz3aGT) (\\"helpers.test.ts\\" 35)) (\\"address.ts\\" 16)) (\\"address.ts\\" 14)) (\\"helpers.test.ts\\" 35)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"helpers.test.ts\\" 32))
        )
        "
`;

exports[`Helpers IsKT1 2`] = `
"{
    \\"micheline\\": \\"parameter (unit %ep1);\\\\nstorage   bool;\\\\ncode\\\\n  {\\\\n    DROP;\\\\n    PUSH address \\\\\\"KT1R9M3MDffw7qSVSnbJs46aMC9YzzZz3aGT\\\\\\";\\\\n    PUSH address \\\\\\"KT1XvNYseNDJJ6Kw27qhSEDF8ys8JhDopzfG\\\\\\";\\\\n    COMPARE;\\\\n    GE;\\\\n    IF\\\\n      {\\\\n        PUSH address \\\\\\"KT1R9M3MDffw7qSVSnbJs46aMC9YzzZz3aGT\\\\\\";\\\\n        PUSH address \\\\\\"KT18amZmM5W7qDWVt2pH6uj7sCEd3kbzLrHT\\\\\\";\\\\n        COMPARE;\\\\n        LE;\\\\n      }\\\\n      {\\\\n        PUSH bool False;\\\\n      };\\\\n    NIL operation;\\\\n    PAIR;\\\\n  };\\",
    \\"json\\": [
        {
            \\"prim\\": \\"storage\\",
            \\"args\\": [
                {
                    \\"prim\\": \\"bool\\"
                }
            ]
        },
        {
            \\"prim\\": \\"parameter\\",
            \\"args\\": [
                {
                    \\"prim\\": \\"unit\\",
                    \\"annots\\": [
                        \\"%ep1\\"
                    ]
                }
            ]
        },
        {
            \\"prim\\": \\"code\\",
            \\"args\\": [
                [
                    {
                        \\"prim\\": \\"DROP\\"
                    },
                    {
                        \\"prim\\": \\"PUSH\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"address\\"
                            },
                            {
                                \\"string\\": \\"KT1R9M3MDffw7qSVSnbJs46aMC9YzzZz3aGT\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"PUSH\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"address\\"
                            },
                            {
                                \\"string\\": \\"KT1XvNYseNDJJ6Kw27qhSEDF8ys8JhDopzfG\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"COMPARE\\"
                    },
                    {
                        \\"prim\\": \\"GE\\"
                    },
                    {
                        \\"prim\\": \\"IF\\",
                        \\"args\\": [
                            [
                                {
                                    \\"prim\\": \\"PUSH\\",
                                    \\"args\\": [
                                        {
                                            \\"prim\\": \\"address\\"
                                        },
                                        {
                                            \\"string\\": \\"KT1R9M3MDffw7qSVSnbJs46aMC9YzzZz3aGT\\"
                                        }
                                    ]
                                },
                                {
                                    \\"prim\\": \\"PUSH\\",
                                    \\"args\\": [
                                        {
                                            \\"prim\\": \\"address\\"
                                        },
                                        {
                                            \\"string\\": \\"KT18amZmM5W7qDWVt2pH6uj7sCEd3kbzLrHT\\"
                                        }
                                    ]
                                },
                                {
                                    \\"prim\\": \\"COMPARE\\"
                                },
                                {
                                    \\"prim\\": \\"LE\\"
                                }
                            ],
                            [
                                {
                                    \\"prim\\": \\"PUSH\\",
                                    \\"args\\": [
                                        {
                                            \\"prim\\": \\"bool\\"
                                        },
                                        {
                                            \\"prim\\": \\"False\\"
                                        }
                                    ]
                                }
                            ]
                        ]
                    },
                    {
                        \\"prim\\": \\"NIL\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"operation\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"PAIR\\"
                    }
                ]
            ]
        }
    ],
    \\"smartpy\\": \\"import smartpy as sp\\\\n\\\\nclass Contract(sp.Contract):\\\\n  def __init__(self):\\\\n    self.init_type(sp.TBool)\\\\n\\\\n  @sp.entry_point\\\\n  def ep1(self, params):\\\\n    sp.set_type(params, sp.TUnit)\\\\n    self.data = (sp.address('KT1XvNYseNDJJ6Kw27qhSEDF8ys8JhDopzfG') >= sp.address('KT1R9M3MDffw7qSVSnbJs46aMC9YzzZz3aGT')) & (sp.address('KT18amZmM5W7qDWVt2pH6uj7sCEd3kbzLrHT') <= sp.address('KT1R9M3MDffw7qSVSnbJs46aMC9YzzZz3aGT'))\\\\n\\\\nsp.add_compilation_target(\\\\\\"visualtez_compilation\\\\\\", Contract())\\"
}"
`;

exports[`Helpers StartsWith 1`] = `
"
        (
            template_id (static_id 0 (\\"helpers.test.ts\\" 10))
            storage ()
            storage_type (\\"bool\\")
            messages ((ep1 True False False True (\\"helpers.test.ts\\" 13) ((set_type (params (\\"helpers.test.ts\\" 13)) \\"unit\\" (\\"helpers.test.ts\\" 18)) (set (data) (call_michelson (op \\"
        DUP;
        SIZE;
        DIG 2;
        SWAP;
        PUSH nat 0;
        SLICE;
        IF_NONE
            {
                DROP;
                PUSH bool False;
            }
            {
                COMPARE;
                EQ;
            };
        \\" \\"string\\" \\"string\\" out \\"bool\\") (\\"string.ts\\" 41) (literal (string \\"Hello\\") (\\"helpers.test.ts\\" 13)) (literal (string \\"He\\") (\\"helpers.test.ts\\" 13))) (\\"helpers.test.ts\\" 13)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"helpers.test.ts\\" 10))
        )
        "
`;

exports[`Helpers StartsWith 2`] = `
"{
    \\"micheline\\": \\"parameter (unit %ep1);\\\\nstorage   bool;\\\\ncode\\\\n  {\\\\n    DROP;\\\\n    PUSH string \\\\\\"He\\\\\\";\\\\n    PUSH string \\\\\\"Hello\\\\\\";\\\\n    DUP;\\\\n    SIZE;\\\\n    DIG 2;\\\\n    SWAP;\\\\n    PUSH nat 0;\\\\n    SLICE;\\\\n    IF_NONE\\\\n      {\\\\n        DROP;\\\\n        PUSH bool False;\\\\n      }\\\\n      {\\\\n        COMPARE;\\\\n        EQ;\\\\n      };\\\\n    NIL operation;\\\\n    PAIR;\\\\n  };\\",
    \\"json\\": [
        {
            \\"prim\\": \\"storage\\",
            \\"args\\": [
                {
                    \\"prim\\": \\"bool\\"
                }
            ]
        },
        {
            \\"prim\\": \\"parameter\\",
            \\"args\\": [
                {
                    \\"prim\\": \\"unit\\",
                    \\"annots\\": [
                        \\"%ep1\\"
                    ]
                }
            ]
        },
        {
            \\"prim\\": \\"code\\",
            \\"args\\": [
                [
                    {
                        \\"prim\\": \\"DROP\\"
                    },
                    {
                        \\"prim\\": \\"PUSH\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"string\\"
                            },
                            {
                                \\"string\\": \\"He\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"PUSH\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"string\\"
                            },
                            {
                                \\"string\\": \\"Hello\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"DUP\\"
                    },
                    {
                        \\"prim\\": \\"SIZE\\"
                    },
                    {
                        \\"prim\\": \\"DIG\\",
                        \\"args\\": [
                            {
                                \\"int\\": \\"2\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"SWAP\\"
                    },
                    {
                        \\"prim\\": \\"PUSH\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"nat\\"
                            },
                            {
                                \\"int\\": \\"0\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"SLICE\\"
                    },
                    {
                        \\"prim\\": \\"IF_NONE\\",
                        \\"args\\": [
                            [
                                {
                                    \\"prim\\": \\"DROP\\"
                                },
                                {
                                    \\"prim\\": \\"PUSH\\",
                                    \\"args\\": [
                                        {
                                            \\"prim\\": \\"bool\\"
                                        },
                                        {
                                            \\"prim\\": \\"False\\"
                                        }
                                    ]
                                }
                            ],
                            [
                                {
                                    \\"prim\\": \\"COMPARE\\"
                                },
                                {
                                    \\"prim\\": \\"EQ\\"
                                }
                            ]
                        ]
                    },
                    {
                        \\"prim\\": \\"NIL\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"operation\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"PAIR\\"
                    }
                ]
            ]
        }
    ],
    \\"smartpy\\": \\"import smartpy as sp\\\\n\\\\nclass Contract(sp.Contract):\\\\n  def __init__(self):\\\\n    self.init_type(sp.TBool)\\\\n\\\\n  @sp.entry_point\\\\n  def ep1(self, params):\\\\n    sp.set_type(params, sp.TUnit)\\\\n    self.data = sp.michelson(\\\\\\"\\\\\\\\n        DUP;\\\\\\\\n        SIZE;\\\\\\\\n        DIG 2;\\\\\\\\n        SWAP;\\\\\\\\n        PUSH nat 0;\\\\\\\\n        SLICE;\\\\\\\\n        IF_NONE\\\\\\\\n            {\\\\\\\\n                DROP;\\\\\\\\n                PUSH bool False;\\\\\\\\n            }\\\\\\\\n            {\\\\\\\\n                COMPARE;\\\\\\\\n                EQ;\\\\\\\\n            };\\\\\\\\n        \\\\\\")('Hello', 'He')\\\\n\\\\nsp.add_compilation_target(\\\\\\"visualtez_compilation\\\\\\", Contract())\\"
}"
`;
